core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java

```diff
@@ -0,0 +1,1341 @@
+            return;
+        }
+
+        boolean needsRounding;
+        if (shift < 64) {
+            needsRounding = roundUp && (low & (1L << (shift - 1))) != 0;
+
+            low = (high << 1 << (63 - shift)) | (low >>> shift);
+            high = high >> shift;
+        }
+        else {
+            needsRounding = roundUp && (high & (1L << (shift - 64 - 1))) != 0;
+
+            low = high >> (shift - 64);
+            high = 0;
+        }
+
+        if (needsRounding) {
+            long tmpHigh = incrementHigh(high, low);
+            long tmpLow = incrementLow(high, low);
+
+            high = tmpHigh;
+            low = tmpLow;
+        }
+
+        result[offset] = high;
+        result[offset + 1] = low;
+    }
+
+    public static Int128 divideRoundUp(long dividendHigh, long dividendLow, int dividendScaleFactor, long divisorHigh, long divisorLow, int divisorScaleFactor)
+    {
+        if (dividendScaleFactor >= MAX_PRECISION) {
+            throwOverflowException();
+        }
+
+        if (divisorScaleFactor >= MAX_PRECISION) {
+            throwOverflowException();
+        }
+
+        long[] quotient = new long[2];
+        long[] remainder = new long[2];
+
+        boolean dividendIsNegative = dividendHigh < 0;
+        boolean divisorIsNegative = divisorHigh < 0;
+        boolean quotientIsNegative = (dividendIsNegative != divisorIsNegative);
+
+        if (dividendIsNegative) {
+            long tmpLow = negateLowExact(dividendHigh, dividendLow);
+            long tmpHigh = negateHighExact(dividendHigh, dividendLow);
+
+            dividendLow = tmpLow;
+            dividendHigh = tmpHigh;
+        }
+
+        if (divisorIsNegative) {
+            long tmpLow = negateLowExact(divisorHigh, divisorLow);
+            long tmpHigh = negateHighExact(divisorHigh, divisorLow);
+
+            divisorLow = tmpLow;
+            divisorHigh = tmpHigh;
+        }
+
+        dividePositives(dividendHigh, dividendLow, dividendScaleFactor, divisorHigh, divisorLow, divisorScaleFactor, quotient, remainder);
+
+        // if (2 * remainder >= divisor) - increment quotient by one
+        shiftLeft(remainder, 1);
+        long remainderLow = remainder[1];
+        long remainderHigh = remainder[0];
+
+        if (compareUnsigned(remainderHigh, remainderLow, divisorHigh, divisorLow) >= 0) {
+            incrementUnsafe(quotient, 0);
+        }
+
+        if (quotientIsNegative) {
+            // negateExact not needed since all positive values can be negated without overflow
+            negate(quotient, 0);
+        }
+
+        return Int128.valueOf(quotient);
+    }
+
+    // visible for testing
+    public static void shiftLeft(long[] decimal, int shift)
+    {
+        long high = decimal[0];
+        long low = decimal[1];
+
+        if (shift < 64) {
+            high = (high << shift) | (low >>> 1 >>> (63 - shift));
+            low = low << shift;
+        }
+        else {
+            high = low << (shift - 64);
```

Comment:
It's pre-existent. I'll change it once I get the new division code in. However, I believe it matters, as it's part of the division algorithm. Otherwise, we'd need to special case those conditions.